hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+    ///
+    /// The valid range for `n` is 2..=15
+    pub fn movup(self, idx: usize) {
+        self.stack.movup(idx);
+        self.asm.push(self.ip, MasmOp::Movup(idx as u8));
+    }
+
+    /// Moves the `n`th word to the top of the stack
+    ///
+    /// The valid range for `n` is 2..=3
+    pub fn movupw(self, idx: usize) {
+        self.stack.movupw(idx);
+        self.asm.push(self.ip, MasmOp::Movupw(idx as u8));
+    }
+
+    /// Moves the element on top of the stack, making it the `n`th element
+    ///
+    /// The valid range for `n` is 2..=15
+    pub fn movdn(self, idx: usize) {
+        self.stack.movdn(idx);
+        self.asm.push(self.ip, MasmOp::Movdn(idx as u8));
+    }
+
+    /// Moves the word on top of the stack, making it the `n`th word
+    ///
+    /// The valid range for `n` is 2..=3
+    pub fn movdnw(self, idx: usize) {
+        self.stack.movdnw(idx);
+        self.asm.push(self.ip, MasmOp::Movdnw(idx as u8));
+    }
+
+    /// Pops a boolean element off the stack, and swaps the top two elements
+    /// on the stack if that boolean is true.
+    ///
+    /// Traps if the conditional is not 0 or 1.
+    pub fn cswap(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::Cswap);
+    }
+
+    /// Pops a boolean element off the stack, and swaps the top two words
+    /// on the stack if that boolean is true.
+    ///
+    /// Traps if the conditional is not 0 or 1.
+    pub fn cswapw(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::Cswapw);
+    }
+
+    /// Pops a boolean element off the stack, and drops the top element on the
+    /// stack if the boolean is true, otherwise it drops the next element down.
+    ///
+    /// Traps if the conditional is not 0 or 1.
+    pub fn cdrop(self) {
+        self.stack.dropn(2);
+        self.asm.push(self.ip, MasmOp::Cdrop);
+    }
+
+    /// Pops a boolean element off the stack, and drops the top word on the
+    /// stack if the boolean is true, otherwise it drops the next word down.
+    ///
+    /// Traps if the conditional is not 0 or 1.
+    pub fn cdropw(self) {
+        self.stack.dropn(5);
+        self.asm.push(self.ip, MasmOp::Cdropw);
+    }
+
+    /// Pops the top element on the stack, and traps if that element is != 1.
+    pub fn assert(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::Assert);
+    }
+
+    /// Pops the top element on the stack, and traps if that element is != 0.
+    pub fn assertz(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::Assertz);
+    }
+
+    /// Pops the top two elements on the stack, and traps if they are not equal.
+    pub fn assert_eq(self) {
+        self.stack.dropn(2);
+        self.asm.push(self.ip, MasmOp::AssertEq);
+    }
+
+    /// Pops the top two words on the stack, and traps if they are not equal.
+    pub fn assert_eqw(self) {
+        self.stack.dropn(8);
+        self.asm.push(self.ip, MasmOp::AssertEq);
+    }
+
```

Comment:
There doesn't seem to be a builder function for the `LocAddr` operation.

Reply:
That's intentional, I don't intend to support accessing locals from inline assembly at this time. If someone wants this, they can use `alloca` to allocate a local in the IR, and then pass the address as an argument to the inline assembly block.